source/extensions/transport_sockets/tls/ocsp/asn1_utility.cc

```diff
@@ -0,0 +1,139 @@
+#include "extensions/transport_sockets/tls/ocsp/asn1_utility.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+namespace Ocsp {
+
+std::string Asn1Utility::cbsToString(CBS& cbs) {
+  auto str_head = reinterpret_cast<const char*>(CBS_data(&cbs));
+  return {str_head, CBS_len(&cbs)};
+}
+
+bool Asn1Utility::isOptionalPresent(CBS& cbs, CBS* data, unsigned tag) {
+  int is_present;
+  if (!CBS_get_optional_asn1(&cbs, data, &is_present, tag)) {
+    throw Envoy::EnvoyException("Failed to parse ASN.1 element tag");
+  }
+
+  return is_present;
+}
+
+std::string Asn1Utility::parseOid(CBS& cbs) {
+  CBS oid;
+  if (!CBS_get_asn1(&cbs, &oid, CBS_ASN1_OBJECT)) {
+    throw Envoy::EnvoyException("Input is not a well-formed ASN.1 OBJECT");
+  }
+  char* oid_text = CBS_asn1_oid_to_text(&oid);
+  if (oid_text == nullptr) {
+    throw Envoy::EnvoyException("Failed to parse oid");
+  }
+
+  std::string oid_text_str(oid_text);
+  OPENSSL_free(oid_text);
+  return oid_text_str;
+}
+
+Envoy::SystemTime Asn1Utility::parseGeneralizedTime(CBS& cbs) {
+  CBS elem;
+  if (!CBS_get_asn1(&cbs, &elem, CBS_ASN1_GENERALIZEDTIME)) {
+    throw Envoy::EnvoyException("Input is not a well-formed ASN.1 GENERALIZEDTIME");
+  }
+
+  auto time_str = cbsToString(elem);
+  // OCSP follows the RFC 5280 enforcement that GENERALIZEDTIME
+  // fields MUST be in UTC, so must be suffixed with a Z character.
+  // Local time or time differential, though a part of the ASN.1
+  // GENERALIZEDTIME spec, are not supported.
+  // Reference: https://tools.ietf.org/html/rfc5280#section-4.1.2.5.2
+  if (time_str.at(time_str.length() - 1) != 'Z') {
```

Comment:
The linked reference uses the `YYYYMMDDHHMMSSZ` format with no explicit mention of casing. I haven't seen an instance where it's used in lower case but the [RFC](https://tools.ietf.org/html/rfc3339#section-5.6) for internet time format and the [ABNF](https://tools.ietf.org/html/rfc2234#section-2.3) notation it uses specify that ascii characters are case-insensitive, so it could just be safer to allow lowercase z.